### ISSUE TITLE
Handle afl_custom_post_process failure gracefully?

### ISSUE THREAD

**Issue by Manouchehri** [AFLplusplus/AFLplusplus#587]
If my afl_custom_post_process fails, is there a reason why returning `0` and setting `out_buf` to null needs to cause a fatal error?

https://github.com/AFLplusplus/AFLplusplus/blob/572944d7267e3612ef8da17a180bc3d8f1a958a7/src/afl-fuzz-run.c#L113-L117

**Comment by domenukk** [comment]
In order to cancel the execution of this testcase?
Hadn't really crossed my mind yet, is this a use case you want to see supported?

**Comment by Manouchehri** [comment]
Yep, that’s correct. At the moment I’m doing a transformation from a binary input to a structure in afl_custom_post_process, and it sometimes fails.

My current workaround is to just do *out_buf = buf and return 0, but that feels wrong (since buf might as well be random bytes, it’s not supposed to be fed directly to my target).

**Comment by vanhauser-thc** [comment]
The purpose of the functionality is that the mutated data is corrected in a way, e.g. setting correct checksum values.
Hence if NULL is returned it means something went horribly wrong and its the fault of the module.
If e.g. a checksum calculation is not possible (e.g. the checksum is at input offset 20, but the input is just 13 bytes long), then it should return the buffer "as-is".
In your special case this likely makes no sense - however it could still be that the buffer as-is could trigger new coverage (e.g. error handling of invalid data).
if we treat this as "skip this input" then a broken postprocess mutation that always return NULL could not be detected.
IMHO the best solution would be that your transformation in fixed so it cant fail. find out why it fails and fix it maybe?
what do you think?

**Comment by Manouchehri** [comment]
> In your special case this likely makes no sense - 

As an example, say afl_custom_post_process has a transformation map like this:

```
0 = AAAAAA
1 = CCECC
2 = EE
3 = DDDD
4 = EEBEE
5 = BBBB
6 = EEE
7 = CC
8 = F
9 = AA
```

And the output is going to be the literal ASCII of the transformation, plus an appended md5sum of the sum of ints.

Also, my fuzz target only accepts a max of 64 chars, so I want to limit my returned size of afl_custom_post_process to that.

Now say we have the following as input:
```
10000004
```

This would result in:
```
CCECCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEBEE + md5sum(1 + 0 + 0 + 0 + 0 + 0 + 0 + 4)
CCECCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEBEE + md5sum(5)
CCECCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEBEEe4da3b7fbbce2345d7772b0674a318d5
```

This ends up being larger than the 64 char limit I wanted. At this point, I don't have a valid buffer (or size) to return. Using `10000004` as the buffer would also be wrong, since I know 100% that it's going to fail with my target, and it would be a waste of CPU cycles.

Example of an input that would be fine:

```
12222224
```

```
CCECCEEEEEEEEEEEEEEBEE70efdf2ec9b086079795c442636b55fb
```

This transformation is fine, since it's only 54 chars.

Basically the majority of transformation are fine, but I expect some edge cases to always fail.

> if we treat this as "skip this input" then a broken postprocess mutation that always return NULL could not be detected.

I think the simplest solution would be to have another environment variable to act as a flag to ignore postprocess mutation errors.

**Comment by vanhauser-thc** [comment]
why dont you cut off as many bytes before transformation so that after transformation it fits 64 bytes?
In my opinion this is the correct behaviour. otherwise you loose the time spent generating that input.

**Comment by Manouchehri** [comment]
In that simple example, yeah it wouldn't be too hard to do. The issue is with more complex situations where there's also a minimum input size.

**Comment by vanhauser-thc** [comment]
then you generate as many randoms as necessary so it has the minimum size?

I am not sure what you are doing is actually a good idea.
because if a new path is found, this will be saved in the queue and selected as input (so CCECCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEEBEEe4da3b7fbbce2345d7772b0674a318d5 in your example).
so your transformation kinda fails in postprocess?

instead you should then write your own afl_custom_fuzz (which also means you dont need afl_custom_postprocess anymore).
then you might need to add afl_custom_queue_new_entry that rewrites the new queue input to the data before transformation.

**Comment by Manouchehri** [comment]
> so your transformation kinda fails in postprocess?

Right, I agree, that's why I don't it to continue in the case it finds a new path. (I think I'm running into this problem already.)

> instead you should then write your own afl_custom_fuzz (which also means you dont need afl_custom_postprocess anymore).
> then you might need to add afl_custom_queue_new_entry that rewrites the new queue input to the data before transformation.

I'm not sure I follow, won't this end being nearly twice as slow? I would have to go through the transformation in afl_custom_fuzz to make sure my input is valid, and then again in afl_custom_queue_new_entry.

**Comment by vanhauser-thc** [comment]
there are several ways to solve it.

1. in afl_custom_fuzz you implement your own mutation with transformation etc. and in afl_custom_queue_new_entry you reverse the transformation.
2. in afl_custom_fuzz you implement your own mutation that mutates in the known structure, not transformation required.
3. you use afl_custom_postprocess to do the transformation and afl_custom_queue_new_entry to reverse the transformation.

**Comment by vanhauser-thc** [comment]
@Manouchehri can we consider this closed or is this still an open item for you?
